apps/plugin-browser/src/pages/popup/popup-contents/action-center/quick-note-action.tsx

```diff
@@ -1,33 +1,125 @@
 
-const createEntityQuery = /* GraphQL */ `
-  mutation createEntity(
-    $entityTypeId: VersionedUrl!
-    $properties: EntityPropertiesObject!
-  ) {
-    createEntity(entityTypeId: $entityTypeId, properties: $properties)
-  }
-`;
-
-const createQuickNote = (text: string) => {
-  return queryApi(createEntityQuery, {
-    entityTypeId:
-      "https://app.hash.ai/@ciaran/types/entity-type/quick-note/v/1",
-    properties: {
-      "https://blockprotocol.org/@blockprotocol/types/property-type/textual-content/":
-        text,
+const quickNoteEntityTypeId =
+  "https://hash.ai/@hash/types/entity-type/quick-note/v/1" as const;
+
+const containsLinkEntityTypeId =
+  "https://hash.ai/@hash/types/entity-type/contains/v/1" as const;
+
+const numericIndexPropertyTypeId =
+  "https://hash.ai/@hash/types/property-type/numeric-index/v/1" as const;
+
+const blockEntityTypeId =
+  "https://hash.ai/@hash/types/entity-type/block/v/1" as const;
+
+const componentIdPropertyTypeId =
+  "https://hash.ai/@hash/types/property-type/component-id/v/1" as const;
+
+const blockDataLinkEntityTypeId =
+  "https://hash.ai/@hash/types/entity-type/block-data/v/1" as const;
+
+const textEntityTypeId =
+  "https://hash.ai/@hash/types/entity-type/text/v/1" as const;
+
+const textualContentPropertyTypeId =
+  "https://blockprotocol.org/@blockprotocol/types/property-type/textual-content/v/2" as const;
+
+const createEntity = (params: {
+  entityTypeId: VersionedUrl;
+  properties: EntityPropertiesObject;
+  linkData?: LinkData;
+}): Promise<Entity> =>
+  queryApi<CreateEntityMutation, CreateEntityMutationVariables>(
+    createEntityMutation,
+    {
+      entityTypeId: params.entityTypeId,
+      properties: params.properties,
+      linkData: params.linkData,
     },
-  }).then(({ data }: { data: { createEntity: Subgraph } }) => {
+  ).then(({ data }) => {
     return data.createEntity;
   });
+
+const createQuickNote = async (text: string) => {
+  const paragraphs = text
+    // Normalize line endings (optional, in case input comes with different OS-specific newlines)
+    .replace(/\r\n/g, "\n")
+    // Split the input by one or more newlines to identify paragraphs
+    .split(/\n+/)
+    // Trim whitespace from each paragraph and filter out any empty strings that might result from excessive newlines
+
+    .map((paragraph) => paragraph.trim())
+    .filter((paragraph) => paragraph.length > 0);
+
+  const [quickNoteEntity, ...blockEntities] = await Promise.all([
+    createEntity({
+      entityTypeId: quickNoteEntityTypeId,
+      properties: {},
+    }),
+    ...paragraphs.map(async (paragraph) => {
+      const [textEntity, blockEntity] = await Promise.all([
+        createEntity({
+          entityTypeId: textEntityTypeId,
+          properties: {
+            [extractBaseUrl(textualContentPropertyTypeId)]: [
+              { tokenType: "text", text: paragraph },
+            ] satisfies TextToken[],
+          },
```

Comment:
Not suggesting you need to do this now, because it relies on regenerating the system types (done in #3563) but with the new fixed-URL system types we can get slightly more type safety by using a type assertion on the entity type we're aiming for, which will warn us if we use an incorrect property for the type, or if the property value is invalid for the property, rather than the current approach which only checks if the value is the value we say it is (not that it matches the property, or that the property matches the entity type).

This relies on the TS compiler detecting insufficient overlap if we use an incorrect property. Unfortunately we can't use `satisfies TextProperties` because the generated types don't have a branded `BaseUrl` as the type of their keys.

```suggestion
          properties: {
            "https://blockprotocol.org/@blockprotocol/types/property-type/textual-content/":
              [{ tokenType: "text", text: paragraph }],
          } as TextProperties,
```

Reply:
I've addressed this as part of https://github.com/hashintel/hash/pull/3536/commits/a6b58cc58b4b92e2afed128edae79921e805fc31